pkg/cmd/auth/login/login.go

```diff
@@ -0,0 +1,284 @@
+
+// TODO should probably use default hostname from mislav's work
+const defaultHostname = "github.com"
+
+type LoginOptions struct {
+	HttpClient func() (*http.Client, error)
+	IO         *iostreams.IOStreams
+	Config     func() (config.Config, error)
+
+	Hostname string
+	Token    string
+}
+
+func NewCmdLogin(f *cmdutil.Factory, runF func(*LoginOptions) error) *cobra.Command {
+	opts := &LoginOptions{
+		HttpClient: f.HttpClient,
+		IO:         f.IOStreams,
+		Config:     f.Config,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "login",
+		Args:  cobra.ExactArgs(0),
+		Short: "Authenticate with a GitHub host",
+		Long: heredoc.Doc(`Authenticate with a GitHub host.
+
+			This interactive command initializes your authentication state either by helping you log into
+			GitHub via browser-based OAuth or by accepting a Personal Access Token.
+
+			The interactivity can be avoided by specifying --with-token and passing a token on STDIN.
+		`),
+		Example: heredoc.Doc(`
+			$ gh auth login
+			# => do an interactive setup
+
+			$ gh auth login --with-token < mytoken.txt
+			# => read token from mytoken.txt and authenticate against github.com
+
+			$ gh auth login --hostname enterprise.internal --with-token < mytoken.txt
+			# => read token from mytoken.txt and authenticate against a GitHub Enterprise instance
+		`),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			isTTY := opts.IO.IsStdinTTY()
+
+			if !isTTY {
+				if !cmd.Flags().Changed("with-token") {
+					return &cmdutil.FlagError{Err: errors.New("--with-token required when not attached to tty")}
+				}
+			}
+
+			wt, _ := cmd.Flags().GetBool("with-token")
+			if wt {
+				if isTTY {
+					return &cmdutil.FlagError{Err: errors.New("expected token on STDIN")}
+				}
+				defer opts.IO.In.Close()
+				token, err := ioutil.ReadAll(opts.IO.In)
+				if err != nil {
+					return &cmdutil.FlagError{Err: fmt.Errorf("failed to read token from STDIN: %w", err)}
+				}
+
+				opts.Token = strings.TrimSpace(string(token))
+				if opts.Hostname == "" {
+					opts.Hostname = defaultHostname
+				}
+			}
+
+			if runF != nil {
+				return runF(opts)
+			}
+
+			return loginRun(opts)
+		},
+	}
+
+	cmd.Flags().StringVar(&opts.Hostname, "hostname", "", "The hostname of the GitHub instance to authenticate with")
+	cmd.Flags().Bool("with-token", false, "If specified, token is read from STDIN")
+
+	return cmd
+}
+
+func loginRun(opts *LoginOptions) error {
+	cfg, err := opts.Config()
+	if err != nil {
+		return err
+	}
+
+	if opts.Token != "" {
+		if opts.Hostname == "" {
+			panic("empty hostname would leak oauth_token")
+		}
+		err := cfg.Set(opts.Hostname, "oauth_token", opts.Token)
+		if err != nil {
+			return err
+		}
+		err = cfg.Write()
+		if err != nil {
+			return err
+		}
+
+		httpClient, err := opts.HttpClient()
+		if err != nil {
+			return err
+		}
+
+		apiClient := api.NewClientFromHTTP(httpClient)
+
+		hasScopes, _, err := apiClient.HasScopes(expectedScopes...)
```

Comment:
Something that just occurred to me: does HasScopes know that `admin:org` satisfies `read:org`? It's something that I had to teach to CheckScopes https://github.com/cli/cli/pull/1359 but I don't think it applied to HasScopes since the latter was only meant for gist back then

Reply:
ah, I don't think it would.

Reply:
oh, uh, `HasScopes` also has `api.github.com` hardcoded :joy: so when I used it to validate tokens for GHE hosts it's just a lie